loader/loader.go

```diff
@@ -0,0 +1,81 @@
+package loader
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/app/internal"
+	"github.com/docker/app/types"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/pkg/errors"
+)
+
+// LoadFromSingleFile loads a docker app from a single-file format (as a reader)
+func LoadFromSingleFile(path string, r io.Reader, ops ...func(*types.App) error) (*types.App, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading single-file")
+	}
+	parts := strings.Split(string(data), types.SingleFileSeparator)
+	if len(parts) != 3 {
+		return nil, errors.Errorf("malformed single-file application: expected 3 documents, got %d", len(parts))
+	}
+	// 0. is metadata
+	metadata := strings.NewReader(parts[0])
+	// 1. is compose
+	compose := strings.NewReader(parts[1])
+	// 2. is settings
+	setting := strings.NewReader(parts[2])
+	appOps := append([]func(*types.App) error{
+		types.WithComposes(compose),
+		types.WithSettings(setting),
+		types.Metadata(metadata),
+	}, ops...)
+	return types.NewApp(path, appOps...)
+}
+
+// LoadFromDirectory loads a docker app from a directory
+func LoadFromDirectory(path string, ops ...func(*types.App) error) (*types.App, error) {
+	appOps := append([]func(*types.App) error{
+		types.MetadataFile(filepath.Join(path, internal.MetadataFileName)),
+		types.WithComposeFiles(filepath.Join(path, internal.ComposeFileName)),
+		types.WithSettingsFiles(filepath.Join(path, internal.SettingsFileName)),
+	}, ops...)
+	return types.NewApp(path, appOps...)
+}
+
+// LoadFromTar loads a docker app from a tarball
+func LoadFromTar(tar string, ops ...func(*types.App) error) (*types.App, error) {
+	f, err := os.Open(tar)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load app from tar")
+	}
+	defer f.Close()
+	appOps := append(ops, types.WithPath(tar))
+	return LoadFromTarReader(f, appOps...)
+}
+
+// LoadFromTarReader loads a docker app from a tarball reader
+func LoadFromTarReader(r io.Reader, ops ...func(*types.App) error) (*types.App, error) {
+	dir, err := ioutil.TempDir("", "load-from-tar")
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load app from tar")
+	}
+	if err := archive.Untar(r, dir, &archive.TarOptions{
+		NoLchown: true,
+	}); err != nil {
+		if err := os.RemoveAll(dir); err != nil {
+			return nil, errors.Wrap(err, "cannot remove temporary folder")
```

Comment:
This error can be misleading, as it shadows the first error. You should Wrap both the errors.